### PR TITLE
Added "synchronize in background" option

### PR DESCRIPTION
**Pull request by bouiboui**

### CODE REVIEW

**Comment by BasileTrujillo** [comment]
Don't forget to add the `execute_in_background` and `bin_path` option into the plugin blueprint and default values ;)

File: git-sync.php [hunk 1]
```diff
@@ -107,12 +107,79 @@ public function init()
...
+
+    /**
+     * Finds the root path of the running Grav instance
+     * @param array $config Git sync plugin config
+     * @return string Grav root path
+     */
+    private static function findRootPath($config = [])
+    {
+        // Find /bin path
+        if (!(\array_key_exists('bin_path', $config)
```
Comment (BasileTrujillo) [inline git-sync.php]: Since this function aim to returns the root path, what about looking for `root_path` instead of `bin_path`?

**Review by BasileTrujillo — CHANGES_REQUESTED** [review]

File: git-sync.php (hunk 1 above)
Comment (bouiboui) [reply git-sync.php]: It's actually looking for the parent folder of the `bin/` folder, which may or may not be root of the Grav instance -- in our case, it's located in the "core" folder that is situated outside the "user" folder.
I'll rename this to `findBinPath` instead, and call `\dirname` in the calling method.

**Review by devsprite — APPROVED** [review]
it seems ok

File: git-sync.php (hunk 1 above)
Comment (bouiboui) [reply git-sync.php]: Changed in the latest commits.

File: blueprints.yaml [hunk 2]
```diff
@@ -125,6 +125,30 @@ form:
...
+        0: No
+      validate:
+        type: bool
+
+    execute_in_background.root_path:
+      type: text
+      label: Absolute path to Grav's root folder
+      description: >
+        Only fill this field if either /bin or /user is not installed in Grav's root folder.<br/>
+        You may need to hit the Save button twice the first time you set this path.
```
Comment (BasileTrujillo) [inline blueprints.yaml]: `If you enabled the git sync trigger on all admin save event,` you may need to hit the Save button twice the first time you set this path.

File: git-sync.php [hunk 3]
```diff
@@ -107,12 +107,92 @@ public function init()
...
     public function synchronize()
     {
         if (!Helper::isGitInstalled() || !Helper::isGitInitialized()) {
             return true;
         }
 
+        // Synchronize in background ?
+        $pluginConfig = $this->config->get('plugins.' . $this->name);
+
+        if (isset($pluginConfig['execute_in_background']['enabled'])
```
Comment (BasileTrujillo) [inline git-sync.php:191]: I think you should also add the following check:
```php
isset($pluginConfig['execute_in_background'])
```
For the case the config file does not have this entry.

File: git-sync.php [hunk 4]
```diff
@@ -107,12 +107,92 @@ public function init()
...
+    {
+        // Find /bin path
+        if (!($binPath = \realpath($path . DIRECTORY_SEPARATOR . 'bin'))) {
+            // /bin path could not be found automatically
+            return false;
+        }
+
+        // Find /plugin executable
+        $pluginPath = $binPath . DIRECTORY_SEPARATOR . 'plugin';
+        return \file_exists($pluginPath) && \is_readable($pluginPath);
```
Comment (BasileTrujillo) [inline git-sync.php:140]: You can also add:
```php
&& \is_executable($pluginPath)
```

**Review by BasileTrujillo — CHANGES_REQUESTED** [review]

File: git-sync.php (hunk 3 above)
Comment (bouiboui) [reply git-sync.php:191]: You don't have to, and it doesn't raise a notice:
image
http://php.net/manual/en/function.isset.php
Something I learned from a common friend back at Exa 😋

File: git-sync.php (hunk 4 above)
Comment (bouiboui) [reply git-sync.php:140]: That's what I did first, but on Windows it returns `false`, and yet it works!

```php
var_dump(is_executable('bin/plugin'));

// bool(false)
```

image

Don't ask me why.

Comment (BasileTrujillo) [reply git-sync.php:140]: 👎  Please uninstall windobe :P

File: git-sync.php (hunk 3 above)
Comment (BasileTrujillo) [reply git-sync.php:191]: 👍

**Review by BasileTrujillo — APPROVED** [review]

**Pull request merged**